### PR TITLE
Start working

### PR DESCRIPTION
**Pull request by sagymauda**
all project comit

### CODE REVIEW

File: src/main/java/com/example/PupilRegistration/DistanceCalcUtils.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package com.example.PupilRegistration;
+
+public class DistanceCalcUtils {
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/DistanceCalcUtils.java:3]: make class final

File: src/main/java/com/example/PupilRegistration/clr/InitData.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.boot.CommandLineRunner;
+import org.springframework.core.annotation.Order;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Component
+@RequiredArgsConstructor
+@Order(1)
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/clr/InitData.java:15]: CommandLineRunner run only when application finished loading, there is no meaning for `@Order` here...

File: src/main/java/com/example/PupilRegistration/clr/InitData.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+public class InitData implements CommandLineRunner {
+
+    @Autowired
+    PupilRepository pupilRepository;
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/clr/InitData.java:19]: add private modifier

File: src/main/java/com/example/PupilRegistration/clr/InitData.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        shahaf.setGpa(400);
+        shahaf.setName("shahf");
+        shahaf.setGrades(gradesShahaf);
+        Pupil gal = new Pupil();
+        gal.setName("gal");
+        gal.setGpa(300);
+        gal.setGrades(gradesGal);
+
+        List<Pupil> pupils = new ArrayList<>();
+        pupils.add(sagi);
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/clr/InitData.java:64]: use addAll, tried to be as much as efficient with lines of code

File: src/main/java/com/example/PupilRegistration/controller/PupilControllerImpl.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+@RestController
+@Log4j2
+public class PupilControllerImpl {
+
+    @Autowired
+    PupilService pupilService;
+
+    @PostMapping(value = "/pupil")
+    public ResponseEntity<Pupil> createPupil(Pupil pupil)throws Exception {
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/controller/PupilControllerImpl.java:19]: add validity check on Pupil

Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/controller/PupilControllerImpl.java:19]: bad practice to declare this function as throws exception... you should handle any known exception

File: src/main/java/com/example/PupilRegistration/data/Grade.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import javax.persistence.*;
+import java.io.Serializable;
+
+@Data
+@Entity
+@Table(name = "grades")
+@NoArgsConstructor
+public class Grade implements Serializable {
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/data/Grade.java:13]: why `implements Serializable` ? do you understand what it does?

File: src/main/java/com/example/PupilRegistration/data/Grade.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @JsonIgnore
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/data/Grade.java:16]: bad practice to use entity object as api, you should create dtos

File: src/main/java/com/example/PupilRegistration/data/School.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @JsonIgnore
+    private Long id;
+
+    private double lat;
+    private double lon;
+    private int minimumGpa;
+    private int maxNumberOfPupils;
+
+    public double getLat() {
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/data/School.java:26]: do you understand what `@Data` is doing?

File: src/main/java/com/example/PupilRegistration/repository/GradeRepository.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package com.example.PupilRegistration.repository;
+
+import com.example.PupilRegistration.data.Grade;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/repository/GradeRepository.java:7]: do we need `@Repository` ?

File: src/main/java/com/example/PupilRegistration/service/PupilService.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+public class PupilService {
+
+    @Autowired
+    PupilRepository pupilRepository;
+
+    public Pupil addPupil(Pupil pupil) throws Exception {
+        try {
+            return pupilRepository.save(pupil);
+        } catch (Exception e) {
+            log.error("cannot save the pupil" + e.getMessage());
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/service/PupilService.java:23]: check out log.error api, it should get to Exception object as an arg

File: src/main/java/com/example/PupilRegistration/service/PupilService.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        }
+        return null;
+    }
+
+    public Boolean setFriendShip(Long pupilA, Long pupilB) throws Exception {
+        Optional<Pupil> askingPupil = pupilRepository.findById(pupilA);
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/service/PupilService.java:29]: why use Optional?

File: src/main/java/com/example/PupilRegistration/service/PupilService.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        Optional<Pupil> receivePupil = pupilRepository.findById(pupilB);
+
+        if (setFriendShipHelper(askingPupil.orElseThrow(Exception::new), receivePupil.orElseThrow(Exception::new))) {
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/service/PupilService.java:32]: why to throw an exception? handle errors if you can

File: src/main/java/com/example/PupilRegistration/service/PupilService.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    private Boolean setFriendShipHelper(Pupil pupilA, Pupil pupilB) {
+        if (isFriend(pupilA,pupilB)) {
+            pupilA.getFriends().add(pupilB);
+            pupilRepository.save(pupilA);
+            return true;
+        }
+        return false;
+    }
+
+    private Boolean isFriend(Pupil pupilA, Pupil pupilB){
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/service/PupilService.java:48]: is this function belong to this service?

File: src/main/java/com/example/PupilRegistration/service/PupilService.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        if (pupilB.getFriends().contains(pupilA)) {
```
Comment (shahaf-sameach) [inline src/main/java/com/example/PupilRegistration/service/PupilService.java:49]: less lines...

**Review by shahaf-sameach — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/PupilRegistration/data/Grade.java (hunk 6 above)
Comment (sagymauda) [reply src/main/java/com/example/PupilRegistration/data/Grade.java:13]: ur right , it uses to pass data between jvms

File: src/main/java/com/example/PupilRegistration/DistanceCalcUtils.java (hunk 1 above)
Comment (sagymauda) [reply src/main/java/com/example/PupilRegistration/DistanceCalcUtils.java:3]: done

File: src/main/java/com/example/PupilRegistration/service/PupilService.java (hunk 11 above)
Comment (sagymauda) [reply src/main/java/com/example/PupilRegistration/service/PupilService.java:29]: this is what findById returns

**Pull request merged**